### ISSUE TITLE
24-7-26 status update

### ISSUE THREAD

**Issue by samuk** [Agroecology-Lab/feldfreund_devkit_ros#104]
<html><head></head><body><h1>Sowbot/Feldfreund: Status Update</h1>
<h2>devkit: Snapshot 52 → 55 | new repo: <code>yolox-disease</code></h2>
<h3>Overview</h3>
<p>Two threads this week: the mission store got a real backend swap (YAML → SQLite by default, YAML kept as a fallback option behind the same interface), and a new standalone repo landed — a YOLOX port of the plant-disease detector, targeting CPU-only inference on the Avaota A1. Also one regression worth flagging in the timestamp-parsing refactor (below).</p>
<hr>
<h3>Mission store: YAML → SQLite (devkit_ui)</h3>
<p><strong><code>missions.py</code> split into <code>MissionYamlStore</code> / <code>MissionSqliteStore</code> behind a shared interface</strong>, selected by file extension. Default <code>MISSIONS_FILE</code> is now <code>/workspace/maps/missions.db</code> (was <code>.yaml</code>) — SQLite is the live backend unless something still points at a <code>.yaml</code> path.</p>
<ul>
<li><strong>SQLite store writes synchronously</strong> (<code>commit()</code> per call) — no queue, no background thread.</li>
<li><strong>YAML store keeps the original async writer thread</strong> (single background writer, debounced, drains to latest snapshot).</li>
<li><strong><code>missions.py</code>'s module docstring is now wrong for the default backend.</strong> It still describes "Persistence is serialised through a single background writer thread" — true for YAML, not true for SQLite, which is now the default. Worth a doc pass so whoever reads this next doesn't assume writes are async.</li>
<li>New <code>actions.py</code> (the <code>ActionDef</code> dataclass + <code>ACTIONS</code> registry, pulled out of <code>missions.py</code>) and <code>time_utils.py</code> (<code>now_utc</code>/<code>now_utc_str</code>/<code>parse_ts</code>, replacing the old <code>_now_utc</code>/<code>_parse_ts</code> module-level helpers) — both now shared by <code>missions.py</code> and <code>obstacles.py</code>.</li>
<li>New <code>test_mission_sqlite_store.py</code> — first unit test coverage this repo has for the mission store. <code>make test</code> added and wired into CI (<code>.github/workflows/test.yml</code>).</li>
</ul>
<blockquote>
<p>⚠️ <strong>Regression in <code>time_utils.parse_ts()</code>.</strong> The old <code>_parse_ts</code> tried ISO-8601 then the legacy <code>dd-mm-yyyy_hh-mm-ss</code> format, catching <code>ValueError</code> on <em>both</em> and returning <code>None</code> if neither matched — matching its documented contract ("Returns None on missing or malformed input"). The new version only catches the first attempt; if the legacy-format parse also fails, the <code>ValueError</code> propagates uncaught. Any genuinely malformed or corrupted timestamp in a stored record will now throw instead of being treated as "never ran." Worth a one-line fix before this hits a field-corrupted YAML/DB record.</p>
</blockquote>
<hr>
<h3>YOLOX disease detector (new repo: <code>yolox-disease</code>)</h3>
<p>Full re-architecture (not a weight conversion) of the existing <code>Plant-Detection-using-YOLOv8</code> Streamlit app, retargeted at the Avaota A1 (Allwinner T527, 8× Cortex-A55, Mali-G56/57, CPU-only inference path).</p>
<ul>
<li><strong>YOLOv8 (ultralytics/torch, AGPL) → YOLOX-Nano (Apache-2.0, NCNN inference, no libtorch on-device)</strong> — driven by RAM/licence constraints on a 1–4GB SBC, not just speed.</li>
<li><strong>No mainline NPU path</strong> — T527's NPU needs Allwinner's closed vendor BSP, which conflicts with the mainline Armbian/U-Boot path already in use. Targets NEON via NCNN on the A55 cores now; NCNN's Vulkan/Mali backend flagged as a possible revisit if the vendor BSP ever becomes viable.</li>
<li><strong><code>setup.sh</code> is idempotent</strong>, handles Python 3.12 + venv, CPU-only torch by default, patches YOLOX's trainer/prefetcher to not assume CUDA, and optionally pulls + converts the Roboflow dataset (YOLO → COCO) in the same pass, syncing <code>CLASS_NAMES</code> to the dataset's real class order.</li>
<li>Full pipeline present: train → export ONNX → export NCNN (with int8 calibration) → <code>deploy/realtime_infer.py</code> (V4L2 capture, decode+NMS in numpy, headless-capable) for the board itself.</li>
<li><strong>Explicitly unvalidated — flagged honestly in the README, not glossed over:</strong> no board, no trained weights, no dataset in this environment. FPS estimate (15–30 fps at 320×320 int8 on comparable SoCs) is stated as a guess, not a measurement.</li>
<li><strong>Class order is asserted, not confirmed</strong> — <code>CLASS_NAMES</code> is carried over from the original app's dict key order, which the original never actually depended on positionally. This port does depend on it positionally (raw NCNN logits). Confirm against real <code>data.yaml</code> before trusting on-device labels.</li>
</ul>
<hr>
<h3>Other changes</h3>
<ul>
<li><strong><code>manage.py</code></strong>: forest3d model mount flipped <code>ro</code> → <code>rw</code>. Combined with <strong><code>topo_to_forest3d.py</code>'s new <code>patch_crop_model_uri()</code> / <code>patch_model_mesh_scale()</code></strong> (new <code>--crop-model</code>, <code>--plant-scale</code>, <code>--scale-category</code> CLI args), the container now writes back into <code>models/crop/&lt;variant&gt;/model.sdf</code> at generation time — patching mesh <code>&lt;scale&gt;</code> directly because Gazebo Sim ignores <code>&lt;include&gt;&lt;scale&gt;</code> (gz-sim#195). Base scale persisted to a new <code>.base_scale</code> file per model so repeated scale calls compose correctly (<code>new = base × factor</code>) instead of compounding.</li>
<li><strong><code>manage.py</code> / <code>topo_to_forest3d.py</code></strong>: <code>--headland</code> now explicitly <code>5.0</code>, tied by comment to <code>row_discovery_node</code>'s worst-case excursion (<code>headland_clearance_m 0.5 + max_search_distance_m 1.5 = 2.0m</code>) — same coupling flagged last week, now landed in code rather than just planned.</li>
<li><strong><code>Dockerfile</code></strong>: <code>EXTERNAL_CACHEBUST</code> bumped 1 → 2, forcing a fresh pull of the external repo layer.</li>
<li><strong><code>Makefile</code></strong>: pylint's target list <strong>dropped <code>devkit_bringup/launch</code></strong> (still covered by <code>ruff</code> separately) — worth confirming that's intentional and not a silent coverage gap. New <code>test</code> target wired into <code>check</code>.</li>
<li><strong><code>obstacles.py</code> / <code>ui_node.py</code></strong>: <code>datetime.now(timezone.utc)</code> → <code>datetime.now(UTC)</code> (Python 3.11+ shorthand) — consistent with last week's <code>datetime.UTC</code> cleanup, not a repeat of the missed-call-sites bug.</li>
<li><strong><code>README.md</code></strong>: new "Rewrite-from-Scratch Cost Estimate" section — ballpark 157,750–370,100 engineering hours (~76–183 person-years, ~$18.9M–$44.4M at $120/hr) to rebuild the stack pulled in by the Dockerfile from scratch, excluding OpenCV/GDAL/Boost/Eigen/PyTorch.</li>
<li><strong><code>AGENTS.md</code></strong>: new <code>skills/</code> directory convention — model-agnostic <code>skills/&lt;name&gt;/SKILL.md</code> docs, read before matching tasks.</li>
<li>New <code>SECURITY.md</code>, <code>maps/.gitignore</code>.</li>
</ul>
<hr>
<h3>Summary</h3>

Area | Change
-- | --
Mission store | YAML → SQLite default backend; both live behind a shared interface
time_utils.parse_ts | Regression — no longer catches malformed legacy-format timestamps, will raise instead of returning None
missions.py docstring | Stale — still describes async writer thread, not true for the new SQLite default
Unit tests | First coverage added (test_mission_sqlite_store.py), wired into CI
yolox-disease | New repo — YOLOX-Nano/NCNN port of the disease detector for Avaota A1, CPU-only, explicitly unvalidated on real hardware
forest3d model mount | ro → rw; enables new mesh-scale/crop-model patching in topo_to_forest3d.py
--headland 5.0 | Landed in code, coupled to row_discovery_node's excursion math
pylint target list | devkit_bringup/launch dropped — confirm intentional
README | New rewrite-cost estimate (~$19–44M) added

<h3>Flags for next check-in</h3>
<ul>
<li>Fix <code>parse_ts()</code> — restore the catch-both-formats behaviour or the malformed-timestamp <code>None</code> contract breaks.</li>
<li>Update <code>missions.py</code>'s module docstring for the SQLite-default write path.</li>
<li>Confirm <code>devkit_bringup/launch</code> was deliberately dropped from pylint's target list.</li>
<li><code>yolox-disease</code> needs real hardware validation — no FPS numbers, no dataset run, no board test yet.</li>
<li>Confirm <code>CLASS_NAMES</code> order against the actual <code>data.yaml</code> before trusting on-device labels.</li>
</ul></body></html>

**Comment by sujandumaru** [comment]
I looked at these flags and thought I could pick up some of these if they're free. I don't think anyone has work on these yet as `parse_ts()` still raises `ValueError` instead of returning `None` for malformed-timestamp. It would be good first issue to work on as I start to learn more about these repos and how to contribute here.
I don't want to duplicate so if anyone already working on this then I can leave these.
